lib/imagewriter/imdimagewriter.cc

```diff
@@ -0,0 +1,430 @@
+	{}
+
+	void writeImage(const Image& image)
+	{
+ 		const Geometry& geometry = image.getGeometry();
+		unsigned numHeads;
+		unsigned numSectors;
+		unsigned numBytes;
+		std::ofstream outputFile(_config.filename(), std::ios::out | std::ios::binary);
+		if (!outputFile.is_open())
+			Error() << "IMD: cannot open output file";
+		unsigned numSectorsinTrack = 0;
+
+		numHeads = geometry.numSides;
+		numSectors = geometry.numSectors;
+		numBytes = geometry.sectorSize;
+
+		Bytes imagenew;
+		ByteWriter bw(imagenew);
+
+		ImdOutputProto::DataRate dataRate = _config.imd().data_rate();
+		if (dataRate == ImdOutputProto::RATE_GUESS)
+		{
+			dataRate = (geometry.numSectors > 10) ? ImdOutputProto::RATE_HD : ImdOutputProto::RATE_DD;
+			if (geometry.sectorSize <= 256)
+				dataRate = ImdOutputProto::RATE_SD;
+			Logger() << fmt::format("IMD: guessing data rate as {}", ImdOutputProto::DataRate_Name(dataRate));
+		}
+
+		ImdOutputProto::RecordingMode recordingMode = _config.imd().recording_mode();
+		if (recordingMode == ImdOutputProto::RECMODE_GUESS)
+		{
+			recordingMode = ImdOutputProto::RECMODE_MFM;
+			Logger() << fmt::format("IMD: guessing recording mode as {}", ImdOutputProto::RecordingMode_Name(recordingMode));
+		}
+
+
+		//Give the user a option to give a comment in the IMD file for archive purposes.
+		std::string comment = _config.imd().comment();
+		if (comment.size() == 0)
+		{
+			comment = LABEL ;
+			comment.append(" date: ");
+			comment.append(__DATE__);
+			comment.append(" time: ");
+			comment.append(__TIME__);
+		} else
+		{
+			comment.insert(0,"IMD ");
+		}
+		bw.seek(0);
+
+		bw.append(comment);
+		bw.write_8(END_OF_FILE);
+		std::string sector_skew;
+		sector_skew.clear();
+		unsigned Status_Sector = 1;
+		bool blnOptionalCylinderMap = false;
+		bool blnOptionalHeadMap = false;
+
+		/* Write the actual sector data. */
+		for (int track = 0; track < geometry.numTracks; track++)
+		{
+			for (int head = 0; head < numHeads; head++)
+			{
+				unsigned sectorIdBase = 1; //IMD starts sector numbering with 1;
+				unsigned sectorId = 0;
+				TrackHeader header = {0, 0, 0, 0, 0}; //define something to hold the header values
+				const auto& sector = image.get(track, head, sectorId+1);
+				if (!sector)  
+				{//sector 0 doesnt exist exit with error
+					//this track, head has no sectors
+					Status_Sector = 0;
+					Logger() << fmt::format("IMD: sector {} not found on track {}, head {}\n", sectorId+1, track, head);
+					break;
+				} else
+				{
+					/* Get the header information */
+					numBytes = sector->data.size();		//number of bytes can change per sector per track
+					header.track = track;
+					header.Head = head;
+					header.SectorSize = setSectorSize(numBytes);
+					sector_skew.clear();
+					numSectorsinTrack = 0;
+				nanoseconds_t RATE = 0;
+				if (sector->clock > 0)
+				{
+					RATE = 1000000.0 / sector->clock;
+				} else
+				{
+					switch (dataRate)
+					{
+					case ImdOutputProto::RATE_HD : RATE = 1000;
```

Comment:
This probably want `break` statements!

Reply:
updates the code with break statements...